### PR TITLE
Whitelists

### PR DESCRIPTION
**Pull request by mhgoldman**

### CODE REVIEW

File: app/assets/stylesheets/subscribers.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the subscribers controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (eliotsykes) [inline app/assets/stylesheets/subscribers.scss]: No need to have these blank assets. Consider creating an initializer to stop the generator from building these:

```
# config/initializers/generators.rb
Rails.application.configure do
  config.generators do |g|
    g.assets false
    g.helper false
  end
end
```

File: app/controllers/self_learners_controller.rb [hunk 2]
```diff
@@ -20,7 +20,7 @@ def create
   def show
     @self_learner = SelfLearner.find_by_id_and_token(params[:id], params[:token])
 
-    unless @self_learner.provisioned?
+    if !@self_learner.provisioned? && !@self_learner.has_running_jobs?
       @self_learner.provision_configuration_later
     end
```
Comment (eliotsykes) [inline app/controllers/self_learners_controller.rb]: `show` action maybe knows more than it needs to about `SelfLearner`-s inner workings.

Consider introducing a method on `SelfLearner` (or modifying `provision_configuration_later`) that wraps lines 23-25 so controller action can be reduced to something like:

```ruby
def show
  @self_learner = SelfLearner.find_by_id_and_token ...
  @self_learner.provision_configuration_later
end
```

File: app/helpers/subscribers_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module SubscribersHelper
```
Comment (eliotsykes) [inline app/helpers/subscribers_helper.rb]: Blank helpers can be safely deleted to help keep app free of clutter.

File: app/controllers/settings_controller.rb [hunk 4]
```diff
@@ -49,7 +48,8 @@ def settings_params
       :student_event_summary_email_text,
       :event_removal_notice_email_text,
       :self_learner_confirmation_email_text,
-      :contact_link_url
+      :contact_link_url,
+      :subscriber_types
```
Comment (eliotsykes) [inline app/controllers/settings_controller.rb]: Check these new settings are tested in spec/features/admin_edits_settings_spec.rb

File: app/controllers/subscribers_controller.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+class SubscribersController < ApplicationController
```
Comment (eliotsykes) [inline app/controllers/subscribers_controller.rb]: Pretty sure this shouldn't be an `AdminController`, but just checking in case.

Comment (eliotsykes) [inline app/controllers/subscribers_controller.rb]: I think there's a feature spec missing for the subscribers functionality.

File: app/models/concerns/has_encrypted_token.rb [hunk 6]
```diff
@@ -2,7 +2,7 @@ module HasEncryptedToken
   extend ActiveSupport::Concern
 
   class_methods do
-    TOKEN_LENGTH = 32
+    TOKEN_LENGTH = Figaro.env.token_length!.to_i
 
     def find_by_id_and_token(id, token)
```
Comment (eliotsykes) [inline app/models/concerns/has_encrypted_token.rb]: Line 10 has a potential timing attack vulnerability due to the comparison here: `token == obj.decrypted_token`.

See https://groups.google.com/forum/#!topic/rubyonrails-security/ANv0HDHEC3k for a fix using `variable_size_secure_compare`

File: config/application.example.yml [hunk 7]
```diff
@@ -29,6 +29,9 @@ schedule_max_concurrency: "5"
 start_at_padding_hours: "1"
 end_at_padding_hours: "1"
 recent_duration_days: "30"
+token_length: "32"
+shadow_users_prefix: "course-manager"
+shadow_users_email: "mgoldman@skytap.com"
```
Comment (eliotsykes) [inline config/application.example.yml]: Consider using a dummy email as the example shadow_users_email.

File: config/routes.rb [hunk 8]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :subscribers
```
Comment (eliotsykes) [inline config/routes.rb]: Whitelist the actions that are available so as not to expose unused routes:
```diff
-  resources :subscribers
+  resources :subscribers, only: [:index, :show, ...etc.]
```

File: spec/controllers/subscribers_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (eliotsykes) [inline spec/controllers/subscribers_controller_spec.rb]: Remove unused skeleton specs.

File: spec/helpers/subscribers_helper_spec.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+# describe SubscribersHelper do
+#   describe "string concat" do
+#     it "concats two strings with spaces" do
+#       expect(helper.concat_strings("this","that")).to eq("this that")
+#     end
+#   end
+# end
+RSpec.describe SubscribersHelper, type: :helper do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (eliotsykes) [inline spec/helpers/subscribers_helper_spec.rb]: Unused spec can safely be deleted to keep app clutter-free

**Comment by eliotsykes** [comment]
@mhgoldman Feedback for PR #66 above

File: app/controllers/subscribers_controller.rb (hunk 5 above)
Comment (mhgoldman) [inline app/controllers/subscribers_controller.rb]: TODO feature spec

**Pull request merged**